fix: dont say reboot required when unnecessary

orndorffgrant opened the pull request:
## Proposed Commit Message
<!-- Include a proposed commit message because all PRs can be merged in a variety of ways by the reviewer -->

> fix: dont say reboot required when unnecessary
> 
> Before, we were assuming that if the system reboot
> flag was set, then it was because of our fix operation.
> Now, we are carving out a case where we can safely know
> that the reboot is not required because of our fix.
> Specifically, if our fix operation didn't actually install
> anything, then it couldn't have caused the reboot to be
> required.
>
> LP: #1926183

## Test Steps
<!-- Please include any steps necessary to verify (and reproduce if
this is a bug fix) this change on a live deployed system,
including any necessary configuration files, user-data,
setup, and teardown. Scripts used may be attached directly to this PR. -->
* Run new unit tests.
* Start a vm and do something that requires reboot. Then run a fix command  that doesn't do anything. Verify it doesn't claim that you need to reboot to complete the fix

## Desired commit type::
<!-- put an `x` in one merge type to allow the reviewer to merge for you -->
 - [x] Squash and merge: Using the "Proposed Commit Message"
 - [ ] Create a merge commit and use "Proposed Commit Message"
 - [ ] Rebase and merge, no merge commit, rely only on existing commit messages

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes
that apply. -->
 - [x] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any integration tests accordingly
 - [ ] I have updated or added any documentation accordingly

lucasmoura commented on uaclient/security.py:
```diff
@@ -822,7 +826,7 @@ def _handle_released_package_fixes(
                 unfixed_pkgs += [src_pkg for src_pkg, _ in pkg_src_group]
 
-    return upgrade_status, unfixed_pkgs
+    return upgrade_status, unfixed_pkgs, all_already_installed
```
I think that's a good approach that you are using. However, we cannot assume that just because we have a binary pkgs to install that they were actually installed.

If you look at `upgrade_packages_and_attach`, if the user is not attached to a ua subscription, we will prompt the user to attach to one to install the package. If the user says no to that, we will not install anything.

lucasmoura commented on uaclient/security.py:
```diff
@@ -813,6 +815,8 @@ def _handle_released_package_fixes(
                         print(status.MESSAGE_SECURITY_UPDATE_INSTALLED)
                         continue
+                    else:
+                        all_already_installed = False
```
I don't think we can just set this variable to `False` here directly. We are looking at three distinct upgrade pockets in this function, `standard updates, esm-apps and esm-infra`. Maybe one of these pockets do not have binary packages to install, but the other do

orndorffgrant replied:
Yes I agree! Note though that this boolean is supposed to only be True when all the binary packages that fix the CVE/USN were already installed.
I'm setting this to False here because in this case, there _are_ packages to be installed, which means that we're not in the special case where they were _all_ already installed.
If we reach this case for at least one of the pockets, then we can't say that "packages were already installed" and the new special case added below won't apply.

lucasmoura replied:
That's fair, but I think that if at least one package was installed, it could require a reboot. That's why I think we should block the reboot message only if no packages were installed at all

orndorffgrant commented:
@lucasmoura I added what we landed on in mattermost - I _think_ this now addresses two problems:
1. Before, if the user had refused to attach, then we might still show the reboot required message.
2. Before, if the fix was already installed, then we might still show the reboot required message.

There are honestly probably more edge cases in here we haven't thought of yet. We may want to come back to this after this release and refactor. But for now I think this addresses the main bugs that people will run into.

Also the various flags were starting to confuse me, so I added a bunch of comments - hopefully they make things more clear, but let me know if you think they're unnecessary.

lucasmoura commented:
@orndorffgrant I completely agree that we should refactor this method. It is getting too bloated and confusing. We should do this in the next release

orndorffgrant commented on uaclient/tests/test_security.py:
```diff
@@ -1522,6 +1522,7 @@ def test_messages_for_affected_packages_when_fix_fail(
         out, err = capsys.readouterr()
         assert expected in out
 
+    @pytest.mark.parametrize("should_reboot", (False, True))
```
@lucasmoura I added this so to ensure this test worked whether should_reboot is False or True

blackboxsw reviewed (commented):
This looks good @orndorffgrant thanks. +1

I think during merge though, we should use squash commit in order to restate your commit message, but we also need to reference this Launchpad bug as the commit mesage footer (to aid in our debian/changelog creation and launchpad closure of the bug once released)

LP: #1926183

blackboxsw commented:
Known CI intermittent failures:
trusty:      ERROR:paramiko.transport:Exception: Error reading SSH protocol banner
bionic:       paramiko.ssh_exception.SSHException: Error reading SSH protocol banner
